### ISSUE TITLE
Non-PSD kernel encountered when computing NTK

### ISSUE THREAD

**Issue by mohamad-amin** [google/neural-tangents#134]
Hey! Thanks for all the amazing work.

I'm trying to compute the NTK for some data on a shallow version of WideResNet, and I'm encountering a non-PSD matrix which results in `NaN`s in predictions resulting from kernel ridge(-less) regression using that NTK, as the computations relies on the cholesky decomposition of the kernel which assumes that the kernel is PSD. The ugly here is that the minimum eigenvalue is too far in the negatives (it's ~ -0.239) that it doesn't seem to be a numerical error (or if it is, I'd be really surprised).

I've provided a simple reproducible Colab notebook that reproduces the example.  The colab needs two files to run properly which are:
* `wide_resnet_jax.py`: The python file containing the shallow WideResNet module implemented using Flax. You can download it from this gist: https://gist.github.com/mohamad-amin/5334109dba81b9c26e7b4d1ded7fd9ad
* `psd_data.pkl`, which can be downloaded from: https://drive.google.com/file/d/18eb93M34vaWjFyNzIq-vnOfll0T6HCjT/view?usp=sharing

`psd_data.pkl` is the pickled version of a dict containing three things:
* `data`: The train and test data used for NTK computations and training the model.
* `params`: The trained parameters of the WideResNet module that we're using, such that it will achieve 1.0 train accuracy and 0.89 test accuracy.
* `labels`: The labels of the datapoints in `data`, to double check the accuracies.

Basically, I would like to have a psd kernel for the train (first 2000 points) section of the data. As the NTK for train data is the outer product of two identical jacobian matrices, thus, theoretically, it must be psd (and PD, as the jacobians would ideally have no linear dependency too, as the datapoints are not linearly dependent).

Possibly helpful note:  I'm using BatchNorm in my network, but have set the `running_mean` parameter of Flax's BatchNorm layers to true, which means it will not compute any running mean/stds using the input data that is passing through the network and it will use the pre-computed parameters. Thus, it should be possible to use `vmap_axes=0`, and I've done so, as the network's behaviour doesn't change among different batches (Ideally, I guess, but it should be true). I haven't tried setting `vmap_axes=None`, but as the network is constant, I'd be confused if it makes any difference.

Colab Link:
https://colab.research.google.com/drive/1I7I1Y0cbOrryNwZUgvuYJsj1gOJWSTix?usp=sharing

**Comment by mohamad-amin** [comment]
Update:
Looks like a flax bug, if my understanding of Flax's batchnorm implementation is correct.

More details:
```
ys = []
for i in range(10):
  ys.append(apply_fn(params, X_train[i:i+1]))
ys = jnp.stack(ys).squeeze()
vs = apply_fn(params, X_train[:10])
np.allclose(ys, vs)
# Outputs False!
```

I don't think this is related to neural-tangents anymore, but I'll leave it open for you to close it if you want or add anything that you might think it could be worth noting.

Thanks!

**Comment by mohamad-amin** [comment]
Update: 

[Seems like](https://github.com/google/flax/issues/1755) the network's output discrepancy only happens when we change the batch size. But when computing this NTK, were not changing the batch-size, are we? So it might still be a neural-tangents bug too.

**Comment by romanngg** [comment]
Thank you for the detailed example and updates! To clarify, the issue goes away when using higher precision? This is a somewhat common issue, I don't have an immediate solution, but can reference some similar issues: https://github.com/google/neural-tangents/issues/128, https://github.com/google/neural-tangents/issues/76, and some discussion in section D of https://arxiv.org/pdf/2007.15801.pdf.
